heron/stmgr/src/cpp/manager/tmaster-client.cpp

```diff
@@ -207,31 +208,38 @@ void TMasterClient::OnHeartbeatTimer() {
 }
 
 void TMasterClient::SendRegisterRequest() {
-  auto request = new proto::tmaster::StMgrRegisterRequest();
-
-  sp_string cwd;
-  FileUtils::getCwd(cwd);
-  proto::system::StMgr* stmgr = request->mutable_stmgr();
-  stmgr->set_id(stmgr_id_);
-  stmgr->set_host_name(stmgr_host_);
-  stmgr->set_data_port(stmgr_port_);
-  stmgr->set_local_endpoint("/unused");
-  stmgr->set_cwd(cwd);
-  stmgr->set_pid((sp_int32)ProcessUtils::getPid());
-  stmgr->set_shell_port(shell_port_);
-  for (auto iter = instances_.begin(); iter != instances_.end(); ++iter) {
-    request->add_instances()->CopyFrom(*(*iter));
-  }
-  SendRequest(request, NULL);
+  CHECK(register_request_set_);
+  proto::tmaster::StMgrRegisterRequest* request =
+    new proto::tmaster::StMgrRegisterRequest(register_request_);
+  SendRequest(request, nullptr);
   return;
 }
 
+void TMasterClient::SetStmgrRegisterRequest(
+                                    const std::vector<proto::system::Instance*>& _instances) {
+    register_request_set_ = true;
+
+    sp_string cwd;
+    FileUtils::getCwd(cwd);
+    proto::system::StMgr* stmgr = register_request_.mutable_stmgr();
+    stmgr->set_id(stmgr_id_);
+    stmgr->set_host_name(stmgr_host_);
+    stmgr->set_data_port(stmgr_port_);
+    stmgr->set_local_endpoint("/unused");
+    stmgr->set_cwd(cwd);
+    stmgr->set_pid((sp_int32)ProcessUtils::getPid());
+    stmgr->set_shell_port(shell_port_);
+    for (auto iter = _instances.begin(); iter != _instances.end(); ++iter) {
+      register_request_.add_instances()->CopyFrom(*(*iter));
+    }
+}
```

Comment:
I belive we could do a simplification here.
Can we rename SetStmgrRegisterRequest to SetInstanceInfo, where the tmaster client copies the instances into local variable. And then during send register, just copies them back into the register request and sends it.

Reply:
The proposed solution is a cleaner way to do the thing. But why is it a simplification? There're two sets of copy operations instead of one. Plus, it requires the management of some objects in the memory.